### PR TITLE
DRILL-7271: Refactor Metadata interfaces and classes to contain all needed information for the File based Metastore

### PR DESCRIPTION
**Pull request by vvysotskyi**
For details please see [DRILL-7271](https://issues.apache.org/jira/browse/DRILL-7271).

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java [hunk 1]
```diff
@@ -361,6 +365,102 @@ public FileMetadata getFileMetadata(Path location) {
...
+        Collections.singletonList(
+            new StatisticsHolder<>(
+                TableStatisticsKind.ROW_COUNT.mergeStatistics(metadataList),
+                TableStatisticsKind.ROW_COUNT));
+    // this code is used only to collect segment metadata to be used only during filtering,
+    // so metadata identifier is not required here and in other places in this class
+    MetadataInfo metadataInfo = new MetadataInfo(MetadataType.SEGMENT, MetadataInfo.GENERAL_INFO_KEY, null);
+    T firstMetadata = metadataList.iterator().next();
+
+    return new SegmentMetadata(firstMetadata.getTableInfo(), metadataInfo, column, firstMetadata.getSchema(),
```
Comment (ihuzenko) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java]: It's 10 arguments constructor. Maybe it's time for builder ? :)

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java [hunk 2]
```diff
@@ -594,50 +594,66 @@ public static long getDateTimeValueFromBinary(Binary binaryTimeStampValue, boole
...
+
+    return Types.optional(minorType);
+  }
+
+  /**
+   * Builds minor type using given {@code OriginalType originalType} or {@code PrimitiveTypeName type}.
+   *
+   * @param type         parquet primitive type
+   * @param originalType parquet original type
+   * @return major type
```
Comment (ihuzenko) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java]:
```suggestion
   * @return minor type
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetTableMetadataUtils.java [hunk 3]
```diff
@@ -148,112 +142,71 @@ private ParquetTableMetadataUtils() {
...
 
-  /**
-   * Returns "natural order" comparator which threads nulls as min values.
-   *
-   * @param <T> type to compare
-   * @return "natural order" comparator
-   */
-  public static <T extends Comparable<T>> Comparator<T> getNaturalNullsFirstComparator() {
-    return Comparator.nullsFirst(Comparator.naturalOrder());
+    return new PartitionMetadata(fileMetadata.getTableInfo(), metadataInfo, partitionColumn, fileMetadata.getSchema(),
```
Comment (ihuzenko) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetTableMetadataUtils.java]: maybe use builder here ?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataProvider.java [hunk 4]
```diff
@@ -71,6 +73,27 @@
    */
   List<FileMetadata> getFilesMetadata();
 
+  /**
+   * Returns list of {@link SegmentMetadata} instances which provides metadata for specific segment and its columns.
+   *
+   * @return list of {@link SegmentMetadata} instances
+   */
+  List<SegmentMetadata> getSegmentsMetadata();
```
Comment (ihuzenko) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataProvider.java]: may be removed in favor of method which returns map

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataProvider.java [hunk 5]
```diff
@@ -71,6 +73,27 @@
    */
   List<FileMetadata> getFilesMetadata();
```
Comment (ihuzenko) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataProvider.java]: may be removed in favor of method which returns map

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataUtils.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                  new StatisticsHolder<>(TableStatisticsKind.ROW_COUNT.getValue(metadata), ColumnStatisticsKind.NULLS_COUNT)));
+        }
+        statisticsList.add(statistics);
+      }
+      List<StatisticsHolder> statisticsHolders = new ArrayList<>();
+      for (CollectableColumnStatisticsKind statisticsKind : statisticsToCollect) {
+        Object mergedStatistic = statisticsKind.mergeStatistics(statisticsList);
+        statisticsHolders.add(new StatisticsHolder<>(mergedStatistic, statisticsKind));
+      }
+      columnsStatistics.put(column, new ColumnStatistics<>(statisticsHolders, statisticsList.iterator().next().getComparatorType()));
```
Comment (ihuzenko) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataUtils.java]: This call to ```statisticsList.iterator().next().getComparatorType()``` is risky, what if somebody will pass empty ```metadataList``` to the method ?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataUtils.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    return columnsStatistics;
+  }
+
+  /**
+   * Updates row cont and column nulls count for specified table metadata and returns new {@link TableMetadata} instance with updated statistics.
```
Comment (ihuzenko) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataUtils.java]:
```suggestion
   * Updates row count and column nulls count for specified table metadata and returns new {@link TableMetadata} instance with updated statistics.
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/Metadata.java [hunk 8]
```diff
@@ -86,4 +88,7 @@
    * @return {@link ColumnMetadata} schema description of the column
    */
   ColumnMetadata getColumn(SchemaPath name);
+
+  TableInfo getTableInfo();
+  MetadataInfo getMetadataInfo();
```
Comment (ihuzenko) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/Metadata.java]: add new line between methods for consistency

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java (hunk 1 above)
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java]: Agree, done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetTableMetadataUtils.java (hunk 3 above)
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetTableMetadataUtils.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetGroupScan.java [hunk 9]
```diff
@@ -281,31 +265,50 @@ public AbstractGroupScanWithMetadata applyFilter(LogicalExpression filterExpr, U
...
       logger.warn("applyFilter {} wasn't able to do pruning for  all metadata levels filter condition, since metadata count for " +
             "{} level exceeds `planner.store.parquet.rowgroup.filter.pushdown.threshold` value.\n" +
             "But underlying metadata was pruned without filter expression according to the metadata with above level.",
-          ExpressionStringBuilder.toString(filterExpr), builder.getOverflowLevel());
+          ExpressionStringBuilder.toString(filterExpr), filteredMetadata.getOverflowLevel());
     }
 
     logger.debug("applyFilter {} reduce row groups # from {} to {}",
-        ExpressionStringBuilder.toString(filterExpr), getRowGroupsMetadata().size(), builder.getRowGroups().size());
+        ExpressionStringBuilder.toString(filterExpr), getRowGroupsMetadata().size(), filteredMetadata.getRowGroups().size());
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetGroupScan.java]: Thanks, done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetGroupScan.java [hunk 10]
```diff
@@ -281,31 +265,54 @@ public AbstractGroupScanWithMetadata applyFilter(LogicalExpression filterExpr, U
...
           .withMatching(false);
     }
 
-    if (builder.getOverflowLevel() != MetadataLevel.NONE) {
-      logger.warn("applyFilter {} wasn't able to do pruning for  all metadata levels filter condition, since metadata count for " +
-            "{} level exceeds `planner.store.parquet.rowgroup.filter.pushdown.threshold` value.\n" +
-            "But underlying metadata was pruned without filter expression according to the metadata with above level.",
-          ExpressionStringBuilder.toString(filterExpr), builder.getOverflowLevel());
+    if (filteredMetadata.getOverflowLevel() != MetadataType.NONE) {
+      if (logger.isWarnEnabled()) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetGroupScan.java]: No objections for this change but what are the odds of warn level being disabled? :)

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java [hunk 11]
```diff
@@ -313,18 +328,30 @@ public TableMetadata getTableMetadata() {
...
+            TableMetadata tableMetadata = getTableMetadata();
+            PartitionMetadata partitionMetadata = PartitionMetadata.builder()
+                .withTableInfo(tableMetadata.getTableInfo())
+                .withMetadataInfo(metadataInfo)
+                .withColumn(partitionColumn)
+                .withSchema(tableMetadata.getSchema())
+                .withColumnsStatistics(columnsStatistics)
+                .withStatistics(statistics)
+                .withPartitionValues(Collections.emptyList())
+                .withLocations((Set<Path>) value)
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java]: Why cast is needed here?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataUtils.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class TableMetadataUtils {
+
+  private TableMetadataUtils() {
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/TableMetadataUtils.java]: Again, no objections but just per my opinion this is an overhead.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataInfo.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+package org.apache.drill.metastore.metadata;
+
+/**
+ * Class which identifies specific metadata.
+ */
+public class MetadataInfo {
+
+  public static final String GENERAL_INFO_KEY = "GENERAL_INFO";
+  public static final String DEFAULT_SEGMENT_KEY = "DEFAULT_SEGMENT";
+  public static final String DEFAULT_COLUMN_PREFIX = "_$SEGMENT_";
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataInfo.java]: Where this constant will be used?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataInfo.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.metastore.metadata;
+
+/**
+ * Class which identifies specific metadata.
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataInfo.java]: Please write better java doc: "Class that specifies metadata type ..." and provide an example.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataType.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+package org.apache.drill.metastore.metadata;
+
+/**
+ * Enum with possible types of metadata.
+ */
+public enum MetadataType {
+
+  ALL,
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataType.java]: java doc: "Metadata that can be applicable to any type"

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataType.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+   * File level metadata type.
+   */
+  FILE,
+
+  /**
+   * Row group level metadata type. Used for parquet tables.
+   */
+  ROW_GROUP,
+
+  NONE
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/MetadataType.java]: 1. Add java doc
2. Where none can be used?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/PartitionMetadata.java [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+   * File locations for this partition
+   *
+   * @return file locations
+   */
+  public Set<Path> getLocations() {
+    return locations;
+  }
+
+  /**
+   * It allows to check the time, when any files were modified. It is in Unix Timestamp
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/PartitionMetadata.java]:
```suggestion
   * Allows to check the time, when any files were modified. It is in Unix Timestamp
```

Comment (arina-ielchiieva) [reply metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/PartitionMetadata.java]: Add timestamp unit of measurement.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/PartitionMetadata.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    return new PartitionMetadataBuilder();
+  }
+
+  public static class PartitionMetadataBuilder extends BaseMetadataBuilder<PartitionMetadataBuilder> {
+    private SchemaPath column;
+    private List<String> partitionValues;
+    private Set<Path> locations;
+    private long lastModifiedTime = BaseTableMetadata.NON_DEFINED_LAST_MODIFIED_TIME;
+
+    public PartitionMetadataBuilder withLocations(Set<Path> locations) {
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/PartitionMetadata.java]: I think you can omit adding with, example: `locations`.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseTableMetadata.java [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+    public BaseTableMetadataBuilder withInterestingColumns(List<SchemaPath> interestingColumns) {
+      this.interestingColumns = interestingColumns;
+      return self();
+    }
+
+    @Override
+    public BaseTableMetadata build() {
+      checkSetValues();
+      Objects.requireNonNull(partitionKeys, "partitionKeys was not set");
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseTableMetadata.java]:
```suggestion
      Objects.requireNonNull(partitionKeys, "partitionKeys were not set");
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+      this.columnsStatistics = columnsStatistics;
+      return self();
+    }
+
+    public T withStatistics(Collection<StatisticsHolder> statistics) {
+      this.statistics = statistics;
+      return self();
+    }
+
+    protected void checkSetValues() {
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java]:
```suggestion
    protected void checkRequiredValues() {
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+      Objects.requireNonNull(tableInfo, "tableInfo was not set");
+      Objects.requireNonNull(metadataInfo, "metadataInfo was not set");
+      Objects.requireNonNull(columnsStatistics, "columnsStatistics was not set");
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java]:
```suggestion
      Objects.requireNonNull(columnsStatistics, "columnsStatistics were not set");
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+      Objects.requireNonNull(statistics, "statistics was not set");
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java]:
```suggestion
      Objects.requireNonNull(statistics, "statistics were not set");
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ * Common provider of tuple schema, column metadata, and statistics for table, partition, file or row group.
+ */
+public abstract class BaseMetadata implements Metadata {
+  protected final TableInfo tableInfo;
+  protected final MetadataInfo metadataInfo;
+  protected final TupleMetadata schema;
+  protected final Map<SchemaPath, ColumnStatistics> columnsStatistics;
+  protected final Map<String, StatisticsHolder> statistics;
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java]: What the difference between statistics and column statistics? Maybe statistics should be named better, for example, generalStatistics or metadataStatistics?
I think for Metastore we used `metadataStatistics` naming ...

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  public static abstract class BaseMetadataBuilder<T extends BaseMetadataBuilder<T>> {
+    protected TableInfo tableInfo;
+    protected MetadataInfo metadataInfo;
+    protected TupleMetadata schema;
+    protected Map<SchemaPath, ColumnStatistics> columnsStatistics;
+    protected Collection<StatisticsHolder> statistics;
+
+    public T withTableInfo(TableInfo tableInfo) {
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseMetadata.java]: Do you think `with` can be removed?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseTableMetadata.java [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+    return new BaseTableMetadataBuilder();
+  }
+
+  public static class BaseTableMetadataBuilder extends BaseMetadataBuilder<BaseTableMetadataBuilder> {
+    private Path location;
+    private long lastModifiedTime = NON_DEFINED_LAST_MODIFIED_TIME;
+    private Map<String, String> partitionKeys;
+    private List<SchemaPath> interestingColumns;
+
+    public BaseTableMetadataBuilder withLocation(Path location) {
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseTableMetadata.java]: Same here.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseTableMetadata.java [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    public BaseTableMetadataBuilder withInterestingColumns(List<SchemaPath> interestingColumns) {
+      this.interestingColumns = interestingColumns;
+      return self();
+    }
+
+    @Override
+    public BaseTableMetadata build() {
+      checkSetValues();
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/BaseTableMetadata.java]: Why not to override this method and call super inside plus add custom checks?

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/TableInfo.java [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+  public static final String UNKNOWN = "UNKNOWN";
+  public static final TableInfo UNKNOWN_TABLE_INFO = new TableInfo(UNKNOWN, UNKNOWN, UNKNOWN, UNKNOWN, UNKNOWN);
+
+  private final String storagePlugin;
+  private final String workspace;
+  private final String name;
+  private final String type;
+  private final String owner;
+
+  public TableInfo(String storagePlugin, String workspace, String name, String type, String owner) {
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/metadata/TableInfo.java]: Make constructor private and add builder.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/statistics/ColumnStatistics.java [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.drill.metastore.statistics.StatisticsHolder.OBJECT_WRITER;
+
+/**
+ * Represents collection of statistics values for specific column.
```
Comment (arina-ielchiieva) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/statistics/ColumnStatistics.java]: Can you please add example.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/statistics/StatisticsHolder.java [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+
+  public StatisticsKind getStatisticsKind() {
+    return statisticsKind;
+  }
+
+  public static StatisticsHolder deserialize(String serialized) throws IOException {
+    return OBJECT_READER.readValue(serialized);
+  }
+
+  public static String serialize(StatisticsHolder statisticsHolder) throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (arina-ielchiieva) [reply metastore/metastore-api/src/main/java/org/apache/drill/metastore/statistics/StatisticsHolder.java]: Please apply the same for other classes.

**Comment by arina-ielchiieva** [comment]
@vvysotskyi thanks for making the changes. +1

**Pull request merged**